New contentrec errors 0.16

## Summary

Seeing some weird new errors on contentrec page when logging in as Learner:

`Unexpected error: 'NoneType' object does not support item assignment `
`TypeError: 'NoneType' object has no attribute '__getitem__'`
## Branch or installer method

Alpha3 deb installer on Ubuntu 0.15

## Snippet from server.log

From 2 different server starts, 
```
[INFO] [2016-02-24 17:31:23,814] cherrypy.access.140456032666064: 127.0.0.1 - - [24/Feb/2016:17:31:23] "GET /static/images/distributed/CC-BY-NC-SA.png HTTP/1.1" 304 - "http://127.0.0.1:8008/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:44.0) Gecko/20100101 Firefox/44.0"
[INFO] [2016-02-24 17:31:23,984] django.request: HTTP Request /securesync/api/user/status/ - Response: 200
[ERROR] [2016-02-24 17:31:24,194] kalite: Error in JSON view: /api/content_recommender
Traceback (most recent call last):
  File "/usr/share/kalite/python-packages/fle_utils/internet/decorators.py", line 26, in api_handle_error_with_json_wrapper_fn
    return handler(request, *args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/kalite/main/api_views.py", line 98, in content_recommender
    resume_recommendations = [set_bool_flag("resume", rec) for rec in get_resume_recommendations(user, request)] if resume else []
  File "/usr/lib/python2.7/dist-packages/kalite/main/api_views.py", line 94, in set_bool_flag
    rec_dict[flag_name] = True
TypeError: 'NoneType' object does not support item assignment
[INFO] [2016-02-24 17:31:24,195] django.request: HTTP Request /api/content_recommender - Response: 500
[INFO] [2016-02-24 17:32:09,053] django.request: HTTP Request / - Response: 200
[INFO] [2016-02-24 17:32:09,117] django.request: HTTP Request / - Response: 200
 
```

```
[INFO] [2016-02-24 18:00:52,861] cherrypy.access.140168277471696: 127.0.0.1 - - [24/Feb/2016:18:00:52] "GET /static/fonts/glyphicons-halflings-regular.woff HTTP/1.1" 200 23292 "http://127.0.0.1:8008/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/48.0.2564.116 Chrome/48.0.2564.116 Safari/537.36"
[ERROR] [2016-02-24 18:00:53,068] kalite: Error in JSON view: /api/content_recommender
Traceback (most recent call last):
  File "/usr/share/kalite/python-packages/fle_utils/internet/decorators.py", line 26, in api_handle_error_with_json_wrapper_fn
    return handler(request, *args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/kalite/main/api_views.py", line 101, in content_recommender
    next_recommendations = [set_bool_flag("next", rec) for rec in get_next_recommendations(user, request)] if next else []
  File "/usr/lib/python2.7/dist-packages/kalite/topic_tools/content_recommendation.py", line 83, in get_next_recommendations
    topic_tree_based_data = get_exercises_from_topics(topic_tree_based_data)
  File "/usr/lib/python2.7/dist-packages/kalite/topic_tools/content_recommendation.py", line 252, in get_exercises_from_topics
    exercises = get_topic_contents(topic_id=topic, kinds=["Exercise"])[:5] #can change this line to allow for more to be returned
TypeError: 'NoneType' object has no attribute '__getitem__'
[INFO] [2016-02-24 18:00:53,152] django.request: HTTP Request /api/content_recommender - Response: 500
[ERROR] [2016-02-24 18:00:54,373] kalite: Error in JSON view: /api/content_recommender
Traceback (most recent call last):
  File "/usr/share/kalite/python-packages/fle_utils/internet/decorators.py", line 26, in api_handle_error_with_json_wrapper_fn
    return handler(request, *args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/kalite/main/api_views.py", line 101, in content_recommender
    next_recommendations = [set_bool_flag("next", rec) for rec in get_next_recommendations(user, request)] if next else []
  File "/usr/lib/python2.7/dist-packages/kalite/topic_tools/content_recommendation.py", line 82, in get_next_recommendations
    topic_tree_based_data = generate_recommendation_data()[current_subtopic]['related_subtopics'][:settings.TOPIC_RECOMMENDATION_DEPTH]
  File "/usr/lib/python2.7/dist-packages/kalite/topic_tools/content_recommendation.py", line 346, in generate_recommendation_data
    other_neighbors = get_subsequent_neighbors(related, recommendation_data, subtopic)
  File "/usr/lib/python2.7/dist-packages/kalite/topic_tools/content_recommendation.py", line 512, in get_subsequent_neighbors
    elif data[left]['related_subtopics'][1].split(' ')[1] == '4': #if the next left neighbor is at dist 4
IndexError: list index out of range
[INFO] [2016-02-24 18:00:54,373] django.request: HTTP Request /api/content_recommender - Response: 500
[INFO] [2016-02-24 18:00:54,654] django.request: HTTP Request /api/content_recommender - Response: 200
```

## How to reproduce

I was unable to consistently reproduce these errors, although they do seem to appear on first login after a restart. Most of the time contentrec seems to load correctly though... Will continue to dig!

## Screenshots

![contentrec](https://cloud.githubusercontent.com/assets/1457929/13293679/f06bc05a-db20-11e5-92cd-fb231d3e9d49.png)

--- a/kalite/distributed/static/js/distributed/topics/views.js
+++ b/kalite/distributed/static/js/distributed/topics/views.js
@@ -236,7 +236,7 @@ var SidebarView = BaseView.extend({
         }
     },
 
-    resize_for_narrow: _.debounce(function() {
+    resize_for_narrow: _.throttle(function() {
         if (this.state_model.get("open")) {
             var current_level = this.state_model.get("current_level");
             var column_width = this.$(".topic-container-inner").width();
@@ -279,7 +279,7 @@ var SidebarView = BaseView.extend({
     }, 100),
 
     // Pretty much the code for pre-back-button sidebar resize
-    resize_for_wide: _.debounce(function() {
+    resize_for_wide: _.throttle(function() {
         var current_level = this.state_model.get("current_level");
         var column_width = this.$(".topic-container-inner").width();
         var last_column_width = 400;

--- a/kalite/distributed/static/js/distributed/topics/views.js
+++ b/kalite/distributed/static/js/distributed/topics/views.js
@@ -285,9 +285,11 @@ var SidebarView = BaseView.extend({
         
         this.width = (current_level-1) * column_width + last_column_width + 10;
         this.sidebar.width(this.width);
-        this.sidebar.css({left: 0});
-        this.sidebarTab.css({left: this.width});
-        
+        if (this.state_model.get("open")) {
+            this.sidebar.css({left: 0});
+            this.sidebarTab.css({left: this.width});
+        }
+
         this.set_sidebar_back();
     }, 100),
 

--- a/kalite/distributed/static/js/distributed/topics/views.js
+++ b/kalite/distributed/static/js/distributed/topics/views.js
@@ -237,44 +237,45 @@ var SidebarView = BaseView.extend({
     },
 
     resize_for_narrow: _.debounce(function() {
-        var current_level = this.state_model.get("current_level");
-        var column_width = this.$(".topic-container-inner").width();
-        var last_column_width = this.$(".topic-container-inner:last-child").width();
-        // Hack to give the last child of .topic-container-inner to be 1.5 times the 
-        // width of their parents. Also, sidebar overflow out of the left side of screen
-        // is computed and set here.
+        if (this.state_model.get("open")) {
+            var current_level = this.state_model.get("current_level");
+            var column_width = this.$(".topic-container-inner").width();
+            var last_column_width = this.$(".topic-container-inner:last-child").width();
+            // Hack to give the last child of .topic-container-inner to be 1.5 times the
+            // width of their parents. Also, sidebar overflow out of the left side of screen
+            // is computed and set here.
 
-        // THE magic variable that controls number of visible panels
-        var numOfPanelsToShow = 4;
+            // THE magic variable that controls number of visible panels
+            var numOfPanelsToShow = 4;
 
-        if ($(window).width() < 1120)
-            numOfPanelsToShow = 3;
+            if ($(window).width() < 1120)
+                numOfPanelsToShow = 3;
 
-        if ($(window).width() < 920)
-            numOfPanelsToShow = 2;
+            if ($(window).width() < 920)
+                numOfPanelsToShow = 2;
 
-        if ($(window).width() < 620)
+            if ($(window).width() < 620)
             numOfPanelsToShow = 1;
+            // Used to get left value in number form
+            var sidebarPanelPosition = this.sidebar.position();
+            var sidebarPanelLeft = sidebarPanelPosition.left;
 
-        // Used to get left value in number form
-        var sidebarPanelPosition = this.sidebar.position();
-        var sidebarPanelLeft = sidebarPanelPosition.left;
+            this.width = (current_level - 1) * column_width + last_column_width + 10;
+            this.sidebar.width(this.width);
+            var sidebarPanelNewLeft = -(column_width * (current_level - numOfPanelsToShow)) + this.sidebarBack.width();
+            if (sidebarPanelNewLeft > 0) sidebarPanelNewLeft = 0;
 
-        this.width = (current_level - 1) * column_width + last_column_width + 10;
-        this.sidebar.width(this.width);
-        var sidebarPanelNewLeft = -(column_width * (current_level - numOfPanelsToShow)) + this.sidebarBack.width();
-        if (sidebarPanelNewLeft > 0) sidebarPanelNewLeft = 0;
+            // Signature color flash (also hides a slight UI glitch)
+            var originalBackColor = this.sidebarBack.css('background-color');
+            this.sidebarBack.css('background-color', this.sidebarTab.css('background-color')).animate({'background-color': originalBackColor});
 
-        // Signature color flash (also hides a slight UI glitch)
-        var originalBackColor = this.sidebarBack.css('background-color');
-        this.sidebarBack.css('background-color', this.sidebarTab.css('background-color')).animate({'background-color': originalBackColor});
-        
-        var self = this;
-        this.sidebar.animate({"left": sidebarPanelNewLeft}, 115, function() {
-            self.set_sidebar_back();
-        });
+            var self = this;
+            this.sidebar.animate({"left": sidebarPanelNewLeft}, 115, function () {
+                self.set_sidebar_back();
+            });
 
-        this.sidebarTab.animate({left: this.sidebar.width() + sidebarPanelNewLeft}, 115);
+            this.sidebarTab.animate({left: this.sidebar.width() + sidebarPanelNewLeft}, 115);
+        }
     }, 100),
 
     // Pretty much the code for pre-back-button sidebar resize
@@ -324,8 +325,6 @@ var SidebarView = BaseView.extend({
             this.sidebarTab.css({left: 0}).html('<span class="icon-circle-right"></span>');
             this.$(".sidebar-fade").hide();
         }
-
-        this.set_sidebar_back();
     }, 100),
 
     set_sidebar_back: function() {

--- a/kalite/topic_tools/content_recommendation.py
+++ b/kalite/topic_tools/content_recommendation.py
@@ -444,7 +444,7 @@ def get_neighbors_at_dist_1(topic_index, subtopic_index, topic):
 
     #else check if there is a neighboring topic (right)
     else:
-        if (topic_index + 1) < len(tree):
+        if (topic_index + 1) < len(tree) and tree[(topic_index+1)]['children']:
             #the 4 denotes the # of nodes in path to this other node, will always be 4
             neighbors.append(tree[(topic_index+1)]['children'][0] + ' 4') 
 

--- a/kalite/topic_tools/content_recommendation.py
+++ b/kalite/topic_tools/content_recommendation.py
@@ -426,12 +426,12 @@ def get_neighbors_at_dist_1(topic_index, subtopic_index, topic):
     next = subtopic_index + 1
 
     #if there is a previous topic (neighbor to left)
-    if(prev > -1 ):
+    if (prev > -1 ) and prev < len(topic['children']):
         neighbors.append(topic['children'][prev] + ' 1') # neighbor on the left side
 
     #else check if there is a neighboring topic (left)    
     else:
-        if (topic_index-1) > -1:
+        if (topic_index-1) > -1 and topic_index - 1 < len(tree) and len(tree[(topic_index-1)]['children']) > 1:
             neighbor_length = len(tree[(topic_index-1)]['children'])
             neighbors.append(tree[(topic_index-1)]['children'][(neighbor_length-1)] + ' 4')
 

--- a/kalite/topic_tools/content_recommendation.py
+++ b/kalite/topic_tools/content_recommendation.py
@@ -36,7 +36,8 @@ def get_resume_recommendations(user, request):
 
     final = get_most_recent_incomplete_item(user)
     if final:
-        return [get_content_item(language=request.language, channel=getattr(final, "channel", "khan"), content_id=final.get("id"))]
+        content = get_content_item(language=request.language, channel=getattr(final, "channel", "khan"), content_id=final.get("id"))
+        return [content] if content else []
     else:
         return []
 
@@ -100,8 +101,9 @@ def get_next_recommendations(user, request):
         if exercise_id in exercise_parents_table:
             subtopic_id = exercise_parents_table[exercise_id]['subtopic_id']
             exercise = get_content_item(language=request.language, content_id=exercise_id)
-            exercise["topic"] = get_content_item(language=request.language, content_id=subtopic_id, topic=True)
-            final.append(exercise)
+            if exercise:
+                exercise["topic"] = get_content_item(language=request.language, content_id=subtopic_id, topic=True) or {}
+                final.append(exercise)
 
 
     #final recommendations are a combination of struggling, group filtering, and topic_tree filtering
@@ -208,8 +210,8 @@ def get_explore_recommendations(user, request):
         if recommended_topic:
 
             final.append({
-                'suggested_topic': get_content_item(language=request.language, content_id=recommended_topic, topic=True),
-                'interest_topic': get_content_item(language=request.language, content_id=subtopic_id, topic=True),
+                'suggested_topic': get_content_item(language=request.language, content_id=recommended_topic, topic=True) or {},
+                'interest_topic': get_content_item(language=request.language, content_id=subtopic_id, topic=True) or {},
             })
 
             added.append(recommended_topic)
@@ -505,14 +507,17 @@ def get_subsequent_neighbors(nearest_neighbors, data, curr):
                     at_four_left = True
 
                 else:
-                    #if immediate left node is 4
-                    if data[ curr ]['related_subtopics'][1].split(' ')[1] == '4': 
-                        at_four_left = True
-                        new_dist = 4
-                    elif data[left]['related_subtopics'][1].split(' ')[1] == '4': #if the next left neighbor is at dist 4
-                        at_four_left = True
-                        new_dist = 4
-                    else: #this means that the next left node is at dist 1
+                    try:
+                        #if immediate left node is 4
+                        if data[ curr ]['related_subtopics'][1].split(' ')[1] == '4':
+                            at_four_left = True
+                            new_dist = 4
+                        elif data[left]['related_subtopics'][1].split(' ')[1] == '4': #if the next left neighbor is at dist 4
+                            at_four_left = True
+                            new_dist = 4
+                        else: #this means that the next left node is at dist 1
+                            new_dist = 1
+                    except IndexError:
                         new_dist = 1
 
                 other_neighbors.append(data[left]['related_subtopics'][1].split(' ')[0] + ' ' + str(new_dist))

--- a/kalite/topic_tools/tests/resume_tests.py
+++ b/kalite/topic_tools/tests/resume_tests.py
@@ -3,76 +3,104 @@ This module contains all tests for the functions invoked to
 get the "Resume" recommendations.
 '''
 import datetime
-
-from kalite.topic_tools.content_recommendation import get_most_recent_incomplete_item
+from kalite.topic_tools.content_recommendation import get_most_recent_incomplete_item, get_resume_recommendations
 from kalite.testing.base import KALiteTestCase
 from kalite.facility.models import Facility, FacilityUser
 from kalite.main.models import ExerciseLog
 
+
 class TestResumeMethods(KALiteTestCase):
+    ORIGINAL_POINTS = 37
+    ORIGINAL_ATTEMPTS = 3
+    ORIGINAL_STREAK_PROGRESS = 20
+    NEW_POINTS_LARGER = 22
+    NEW_ATTEMPTS = 5
+    NEW_STREAK_PROGRESS_LARGER = 10
+    NEW_POINTS_SMALLER = 0
+    NEW_STREAK_PROGRESS_SMALLER = 0
+    EXERCISE_ID = "number_line"
+    EXERCISE_ID2 = "radius_diameter_and_circumference"
+    INVALID_EXERCISE_ID = "s_diameter_and_circumference"
+    USERNAME1 = "test_user_resume1"
+    USERNAME2 = "test_user_resume2"
+    USERNAME3 = "test_user_resume3"
+    PASSWORD = "dummies"
+    FACILITY = "Test Facility Resume"
+    TIMESTAMP_LATER = datetime.datetime(2014, 11, 17, 20, 51, 2, 342662)
+    TIMESTAMP_EARLY = datetime.datetime(2014, 10, 8, 15, 59, 59, 370290)
 
-	ORIGINAL_POINTS = 37
-	ORIGINAL_ATTEMPTS = 3
-	ORIGINAL_STREAK_PROGRESS = 20
-	NEW_POINTS_LARGER = 22
-	NEW_ATTEMPTS = 5
-	NEW_STREAK_PROGRESS_LARGER = 10
-	NEW_POINTS_SMALLER = 0
-	NEW_STREAK_PROGRESS_SMALLER = 0
-	EXERCISE_ID = "number_line"
-	EXERCISE_ID2 = "radius_diameter_and_circumference"
-	USERNAME1 = "test_user_resume1"
-	USERNAME2 = "test_user_resume2"
-	PASSWORD = "dummies"
-	FACILITY = "Test Facility Resume"
-	TIMESTAMP_LATER = datetime.datetime(2014, 11, 17, 20, 51, 2, 342662)
-	TIMESTAMP_EARLY = datetime.datetime(2014, 10, 8, 15, 59, 59, 370290)
+    def setUp(self):
+        """Performed before every test"""
 
-	def setUp(self):
-		'''Performed before every test'''
+        # a brand new user
+        self.facility = Facility(name=self.FACILITY)
+        self.facility.save()
 
-		#a brand new user 
-		self.facility = Facility(name=self.FACILITY)
-		self.facility.save()
+        self.user_with_no_activity = FacilityUser(username=self.USERNAME1, facility=self.facility)
+        self.user_with_no_activity.set_password(self.PASSWORD)
+        self.user_with_no_activity.save()
 
-		self.user_with_no_activity = FacilityUser(username=self.USERNAME1, facility=self.facility)
-		self.user_with_no_activity.set_password(self.PASSWORD)
-		self.user_with_no_activity.save()
+        # a user with valid exercises
+        self.user_with_activity = FacilityUser(username=self.USERNAME2, facility=self.facility)
+        self.user_with_activity.set_password(self.PASSWORD)
+        self.user_with_activity.save()
 
-		#a user with valid exercises
-		self.user_with_activity = FacilityUser(username=self.USERNAME2, facility=self.facility)
-		self.user_with_activity.set_password(self.PASSWORD)		
-		self.user_with_activity.save()							
+        # a user with invalid exercises
+        self.user_with_old_activity = FacilityUser(username=self.USERNAME3, facility=self.facility)
+        self.user_with_old_activity.set_password(self.PASSWORD)
+        self.user_with_old_activity.save()
 
-		#add some exercises for second user (both incomplete)
-		self.original_exerciselog2 = ExerciseLog(exercise_id=self.EXERCISE_ID, user = self.user_with_activity, complete=False)
-		self.original_exerciselog2.points = self.ORIGINAL_POINTS
-		self.original_exerciselog2.attempts = self.ORIGINAL_POINTS
-		self.original_exerciselog2.streak_progress = self.ORIGINAL_STREAK_PROGRESS
-		self.original_exerciselog2.latest_activity_timestamp = self.TIMESTAMP_EARLY
-		self.original_exerciselog2.completion_timestamp = self.TIMESTAMP_EARLY
-		self.original_exerciselog2.save()
+        # add some exercises for second user (both incomplete)
+        self.original_exerciselog2 = ExerciseLog(exercise_id=self.EXERCISE_ID, user=self.user_with_activity,
+                                                 complete=False)
+        self.original_exerciselog2.points = self.ORIGINAL_POINTS
+        self.original_exerciselog2.attempts = self.ORIGINAL_POINTS
+        self.original_exerciselog2.streak_progress = self.ORIGINAL_STREAK_PROGRESS
+        self.original_exerciselog2.latest_activity_timestamp = self.TIMESTAMP_EARLY
+        self.original_exerciselog2.completion_timestamp = self.TIMESTAMP_EARLY
+        self.original_exerciselog2.save()
 
-		self.original_exerciselog2 = ExerciseLog(exercise_id=self.EXERCISE_ID2, user = self.user_with_activity, complete=False)
-		self.original_exerciselog2.points = self.ORIGINAL_POINTS
-		self.original_exerciselog2.attempts = self.ORIGINAL_POINTS
-		self.original_exerciselog2.streak_progress = self.ORIGINAL_STREAK_PROGRESS
-		self.original_exerciselog2.latest_activity_timestamp = self.TIMESTAMP_LATER
-		self.original_exerciselog2.completion_timestamp = self.TIMESTAMP_LATER
-		self.original_exerciselog2.save()		
+        self.original_exerciselog2 = ExerciseLog(exercise_id=self.EXERCISE_ID2, user=self.user_with_activity,
+                                                 complete=False)
+        self.original_exerciselog2.points = self.ORIGINAL_POINTS
+        self.original_exerciselog2.attempts = self.ORIGINAL_POINTS
+        self.original_exerciselog2.streak_progress = self.ORIGINAL_STREAK_PROGRESS
+        self.original_exerciselog2.latest_activity_timestamp = self.TIMESTAMP_LATER
+        self.original_exerciselog2.completion_timestamp = self.TIMESTAMP_LATER
+        self.original_exerciselog2.save()
 
-	def test_get_most_recent_incomplete_item(self):
-		'''get_most_recent_incomplete_item()'''
+        self.original_exerciselog3 = ExerciseLog(exercise_id=self.INVALID_EXERCISE_ID, user=self.user_with_old_activity,
+                                                 complete=False)
+        self.original_exerciselog3.points = self.ORIGINAL_POINTS
+        self.original_exerciselog3.attempts = self.ORIGINAL_POINTS
+        self.original_exerciselog3.streak_progress = self.ORIGINAL_STREAK_PROGRESS
+        self.original_exerciselog3.latest_activity_timestamp = self.TIMESTAMP_LATER
+        self.original_exerciselog3.completion_timestamp = self.TIMESTAMP_LATER
+        self.original_exerciselog3.save()
 
-		#test user with activity first
-		expected = {
-			"id": unicode(self.EXERCISE_ID2, 'utf-8'),
-			"timestamp": self.TIMESTAMP_LATER,
-			"kind": "Exercise"
-		}
-		actual = get_most_recent_incomplete_item(self.user_with_activity)
-		self.assertEqual(expected, actual)
+    def test_get_most_recent_incomplete_item(self):
+        '''get_most_recent_incomplete_item()'''
 
+        # test user with activity first
+        expected = {
+            "id": unicode(self.EXERCISE_ID2, 'utf-8'),
+            "timestamp": self.TIMESTAMP_LATER,
+            "kind": "Exercise"
+        }
+        actual = get_most_recent_incomplete_item(self.user_with_activity)
+        self.assertEqual(expected, actual)
 
-		#new user just created (no activity logged)
-		self.assertIsNone(get_most_recent_incomplete_item(user=self.user_with_no_activity))
+        # new user just created (no activity logged)
+        self.assertIsNone(get_most_recent_incomplete_item(user=self.user_with_no_activity))
+
+    def test_get_empty_list_invalid_resume(self):
+        # Used to mock a request object that is only queried for its 'lang' property.
+        class MicroMock(object):
+            def __init__(self, **kwargs):
+                self.__dict__.update(kwargs)
+
+        # test user with invalid activity
+        actual = get_resume_recommendations(self.user_with_old_activity, MicroMock(language="en"))
+
+        # ensure that no recommendations are returned
+        self.assertEqual(len(actual), 0)

